stl/inc/vector

```diff
@@ -458,6 +494,139 @@ public:
     using difference_type = typename _Alty_traits::difference_type;
 
 private:
+#ifdef _INSERT_VECTOR_ANNOTATION
+    static _CONSTEXPR20 void _Create_annotation(const vector& _Myvec_) noexcept {
+        // Annotates the shadow memory of the valid range
+        auto& _My_data = _Myvec_._Mypair._Myval2;
+        _Apply_annotation(_My_data._Myfirst, _My_data._Myend, _My_data._Myend, _My_data._Mylast);
+    }
+
+    static _CONSTEXPR20 void _Clear_annotation(const vector& _Myvec_) noexcept {
+        // Removes the shadow memory annotation of the range
+        auto& _My_data = _Myvec_._Mypair._Myval2;
+        _Apply_annotation(_My_data._Myfirst, _My_data._Myend, _My_data._Mylast, _My_data._Myend);
+    }
+
+    static _CONSTEXPR20 void _Change_annotation(const vector& _Myvec_, const difference_type _Count) noexcept {
+        // Extends /shrinks the annotated range by _Count
+        auto& _My_data = _Myvec_._Mypair._Myval2;
+        _Apply_annotation(_My_data._Myfirst, _My_data._Myend, _My_data._Mylast, _My_data._Mylast + _Count);
+    }
+
+    static const void* _Get_aligned_first(const void* _First, const void* _End) noexcept {
+        const char* _CFirst    = reinterpret_cast<const char*>(_First);
+        const char* _CEnd      = reinterpret_cast<const char*>(_End);
+        const size_t _Capacity = static_cast<size_t>(_CEnd - _CFirst);
+
+        if (_Capacity >= 8) {
+            // We are guaranteed to have sufficient space to find an aligned address.
+            return reinterpret_cast<void*>((reinterpret_cast<uintptr_t>(_CFirst) + 7) & ~7);
+        }
+
+        uintptr_t _Alignment_offset = reinterpret_cast<uintptr_t>(_CFirst) & 7;
+        if (_Alignment_offset) {
+            _Alignment_offset = 8 - _Alignment_offset;
+        }
+
+        if (_Capacity > _Alignment_offset) {
+            return _CFirst + _Alignment_offset;
+        }
+
+        return nullptr;
+    }
+
+    static _CONSTEXPR20 void _Apply_annotation(
+        pointer _First_, pointer _End_, pointer _Old_last_, pointer _New_last_) noexcept {
+#if _HAS_CXX20
+        if (_STD is_constant_evaluated()) {
+            return;
+        }
+#endif // _HAS_CXX20
+
+        if (!_Asan_vector_should_annotate) {
+            return;
+        }
+
+        const char* _First    = reinterpret_cast<const char*>(_Unfancy_maybe_null(_First_));
+        const char* _End      = reinterpret_cast<const char*>(_Unfancy_maybe_null(_End_));
+        const char* _Old_last = reinterpret_cast<const char*>(_Unfancy_maybe_null(_Old_last_));
+        const char* _New_last = reinterpret_cast<const char*>(_Unfancy_maybe_null(_New_last_));

```

Comment:
It appears that these could be top-level `const`. Additionally, as they've just been `reinterpret_cast`ed, we can avoid repeating that, and say `const auto`.